### PR TITLE
[NFC] Enforce use of `Type::` on type names

### PR DESCRIPTION
**Pull request by tlively**
There's some test failure in this change even though it is supposed to be nonfunctional. But before I fix that I thought I would see if this is even a change people are interested in making.

Please don't spend time going through this methodically right now. Just initial impressions would be good.

### CODE REVIEW

**Comment by aheejin** [comment]
I prefer this over #2433. Enforcing use of `Type::` is clearer and has less risk of clashing, and we don't need something like `constexpr Type none = Type::none;` anymore.

**Comment by kripken** [comment]
How does this related to #2433? Do we need just one of the two?

File: src/wasm-type.h [hunk 1]
```diff
@@ -21,19 +21,46 @@
...
-  f32,
-  f64,
-  v128,
-  anyref,
-  exnref,
-  // none means no type, e.g. a block can have no return type. but unreachable
-  // is different, as it can be "ignored" when doing type checking across
-  // branches
-  unreachable
+class Type {
```
Comment (sbc100) [inline src/wasm-type.h]: Wouldn't just switching to `enum class` get you most of the benefits?

**Comment by tlively** [comment]
This PR builds on #2433 and includes it as the first commit. It's a shame that GitHub doesn't support stacked PRs. @sbc100 `enum class` wouldn't work because we need to to be a class to support multivalue, see #2433.

**Comment by sbc100** [comment]
Perhaps switching to enum class first, which would keep 99% of this change the same might work then?   That was you can do a followup proposal to make it into an actual class it might be easier to review (since it would eliminate a lot of the noise from the this change)?  Just a thought.

**Comment by tlively** [comment]
I was hoping to get #2433 merged first, which is a small PR that should be easy to review. Then I would rebase this PR on top of that.

**Comment by sbc100** [comment]
Ah yes, that makes sense.  Sorry for the noise.

**Comment by tlively** [comment]
I still have to go through this and reformat comments, but when that is done this will be ready for review.

**Comment by tlively** [comment]
Now that all the large-scale refactorings have settled down, I believe this is a good time to get this change in @kripken @aheejin.

File: src/wasm-type.h [hunk 2]
```diff
@@ -148,18 +148,6 @@ std::ostream& operator<<(std::ostream& os, ParamType t);
...
-constexpr Type i64 = Type::i64;
-constexpr Type f32 = Type::f32;
-constexpr Type f64 = Type::f64;
-constexpr Type v128 = Type::v128;
-constexpr Type funcref = Type::funcref;
-constexpr Type anyref = Type::anyref;
-constexpr Type nullref = Type::nullref;
-constexpr Type exnref = Type::exnref;
-constexpr Type unreachable = Type::unreachable;
-
```
Comment (kripken) [inline src/wasm-type.h]: this enforcees that we must use the `Type::` notation, is that right?

Comment (tlively) [reply src/wasm-type.h]: That's correct. I'll follow up separately on #2572; it isn't materially affected by this PR.

**Pull request merged**